### PR TITLE
[driver] RW layer and image commit

### PR DESCRIPTION
**Pull request by caervs**
Re-implement the driver interface to mount the volume with a RW layer and
commit/push the layer when the volume is removed. For the time being we can use
a hack where we create a container from the image and bind-mount the graph
driver's merged directory into the target container. Later on the docker engine
may provide the necessary primitives to union the layers and give us a
commitable rw layer.

### CODE REVIEW

File: glide.lock [hunk 1]
```diff
@@ -0,0 +1,64 @@
+hash: 6b4ef86ba3b26a4ff92055f8700395aef07af84c7f97df8642cb8f6459097036
```
Comment (pchico83) [inline glide.lock:1]: are you committing this file?

Comment (vdemeester) [reply glide.lock:1]: @pchico83 `glide.lock` is meant to be commited yes — it makes the build reproductible and "freeze" dependency version (I don't like `glide` very much but that's another *personal* store 😂 👼 ).

**Comment by vdemeester** [comment]
Something I wonder using container as "volume" here, is that they will show in `docker ps` and other related commands. We might want to use containerd grpc API instead so that these containers are invisible (but we loose some graphdriver ability I think… tricky… thinking outloud 🤔)

**Comment by pchico83** [comment]
@caervs @fermayo what is the plan to merge the images generated by two pipeline steps executed in parallel?

**Comment by caervs** [comment]
@pchico83 great question. We don't have a solution yet. For the time being maybe we can mount both images at different paths in the downstream container.

Ideally we actually merge them into a single image and stop the pipeline if there are conflicts. That's not possible with the container hack but maybe containerd/docker can start to give us primitives for creating and merging file systems from images and vice versa.

@vdemeester yes, let's see if containerd gives the same graphdriver ability so that the containers don't show up in docker.

**Comment by fermayo** [comment]
That's a good point @vdemeester, for now everything is a hack anyway - we have to at some point think carefully about what will it take to make all this a "proper" feature (which will include adding primitives to docker engine that we don't have today)

File: driver.go [hunk 2]
```diff
@@ -2,88 +2,180 @@ package main
...
+	if container.GraphDriver.Name != "overlay" {
+		return volume.Response{Err: fmt.Sprintf("unexpected graph driver: %s", container.GraphDriver.Name)}
+	}
+	mergedDir, ok := container.GraphDriver.Data["MergedDir"]
+	if !ok {
+		return volume.Response{Err: fmt.Sprintf("missing MergedDir")}
+	}
+	// HACK directory is relative to host but docker will prepend rootfs of the
+	// plugin container
+	mergedDir = fmt.Sprintf("../../../../../../../../../../../%s", mergedDir)
```
Comment (fermayo) [inline driver.go:114]: This won't work with the `docker service create` way of launching the plugin. Can you change it assuming that the mountpoint is relative to `/`, or make it compatible with both ways of deploying it? (Which might be possible using `PropagatedMount`, I haven't looked in detail)

**Pull request closed without merging**